dbmgr/dbmgr.go

```diff
@@ -0,0 +1,985 @@
+	res := q.DB.Delete(&User{})
+	if err := res.Error; err != nil {
+		return err
+	}
+
+	if res.RowsAffected == 0 {
+		return gorm.ErrRecordNotFound
+	}
+
+	return nil
+}
+
+// AUTH TOKENS
+
+type AuthTokensQuery struct{ DB *gorm.DB }
+type AuthToken struct {
+	gorm.Model
+	Token  string `gorm:"unique"`
+	User   UserID
+	Expiry time.Time
+}
+
+func NewAuthTokensQuery(db *gorm.DB) *AuthTokensQuery {
+	return &AuthTokensQuery{DB: db.Model(AuthToken{})}
+}
+
+func (q *AuthTokensQuery) Create(authToken AuthToken) error {
+	return q.DB.Create(&authToken).Error
+}
+
+// CONTENTS
+
+type ContentsQuery struct{ DB *gorm.DB }
+type ContentID uint
+type Content struct {
+	ID        uint `gorm:"primarykey"`
+	CreatedAt time.Time
+	UpdatedAt time.Time
+	DeletedAt gorm.DeletedAt `gorm:"index"`
+
+	Cid         util.DbCID
+	Name        string
+	UserID      UserID
+	Description string
+	Size        int64
+	Active      bool
+	Offloaded   bool
+	Replication int
+
+	AggregatedIn ContentID
+	Aggregate    bool
+
+	Pinning bool
+	PinMeta string
+
+	Failed bool
+
+	Location string
+}
+
+func NewContentsQuery(db *gorm.DB) *ContentsQuery {
+	return &ContentsQuery{DB: db.Model(Content{})}
+}
+
+func (q *ContentsQuery) WithID(id uint) *ContentsQuery {
+	q.DB = q.DB.Where("id = ?", id)
+	return q
+}
+
+func (q *ContentsQuery) WithActive(active bool) *ContentsQuery {
+	if active {
+		q.DB = q.DB.Where("active")
+	} else {
+		q.DB = q.DB.Where("NOT active")
+	}
+	return q
+}
+
+func (q *ContentsQuery) WithUserID(userID uint) *ContentsQuery {
+	q.DB = q.DB.Where("user_id = ?", userID)
+	return q
+}
+
+func (q *ContentsQuery) WithCid(cid gocid.Cid) *ContentsQuery {
+	q.DB = q.DB.Where("cid = ?", cidToBytes(cid))
+	return q
+}
+
+func (q *ContentsQuery) WithCids(cids []gocid.Cid) *ContentsQuery {
+	q.DB = q.DB.Where("cid IN ?", cidsToBytes(cids))
+	return q
+}
+
+func (q *ContentsQuery) WithAggregate(aggregate bool) *ContentsQuery {
+	if aggregate {
+		q.DB = q.DB.Where("aggregate")
+	} else {
+		q.DB = q.DB.Where("NOT aggregate")
+	}
+	return q
+}
+
+func (q *ContentsQuery) WithAggregatedIn(contentID uint) *ContentsQuery {
+	q.DB = q.DB.Where("aggregated_in = ?", contentID)
+	return q
+}
+
+func (q *ContentsQuery) Limit(limit int) *ContentsQuery {
+	q.DB = q.DB.Limit(limit)
+	return q
+}
+
+func (q *ContentsQuery) Offset(offset int) *ContentsQuery {
+	q.DB = q.DB.Offset(offset)
+	return q
+}
+
+// TODO: order functions can probably be simplified
+func (q *ContentsQuery) OrderByCreationDate(order DBSortOrder) *ContentsQuery {
+	if order == OrderDescending {
+		q.DB = q.DB.Order("created_at DESC")
+	} else {
+		q.DB = q.DB.Order("created_at ASC")
+	}
+	return q
+}
+
+func (q *ContentsQuery) OrderByID(order DBSortOrder) *ContentsQuery {
+	if order == OrderDescending {
+		q.DB = q.DB.Order("id DESC")
+	} else {
+		q.DB = q.DB.Order("id ASC")
+	}
+	return q
+}
+
+func (q *ContentsQuery) CreateAll(contents []Content) error {
+	return q.DB.Create(&contents).Error
```

Comment:
passing a pointer to an array here doesnt seem quite right. I dont think that will do what we expect

Reply:
my mistake, will fix. although we don't actually even create multiple contents anywhere in the program anyway, right? not sure why i added this.